boards: raspberrypi: rpi_pico2w: Add Pico 2W

MagpieEmbedded opened the pull request:
The Pico 2W board is similar to the Pico 2 apart from the addition of an Infineon Wifi chip.
This commit is based on the Pico and Pico W setup. There is a small change to the pin definition to allow communication with the Wifi chip using the Pio SPI device.

Currently dependent on this PR to fix the AIROC configuration: https://github.com/zephyrproject-rtos/zephyr/pull/90789

MagpieEmbedded committed:
boards: raspberrypi: rpi_pico2w: Add Pico 2W

The Pico 2W board is similar to the Pico 2 apart from the addition of an
Infineon Wifi chip.
This commit is based on the Pico and Pico W setup.
A different sample application is shown as blinky is not yet supported
on the Pico 2W.

Signed-off-by: Magpie Embedded <magpieembedded@gmail.com>

nordicjm commented on boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w_defconfig:
```diff
@@ -0,0 +1,17 @@
+CONFIG_GPIO=y
+CONFIG_RESET=y
+CONFIG_SERIAL=y
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=150000000
```
@soburi this needs to be set in soc Kconfig from dts property

soburi replied:
https://github.com/zephyrproject-rtos/zephyr/pull/90893 has been merged.

MagpieEmbedded replied:
Should be updated as such now. I've checked the basic functionality still works for me (when including the #90789 change.

MagpieEmbedded commented on boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w.dts:
```diff
@@ -0,0 +1,79 @@
+	airoc_wifi_host_wake: airoc_wifi_host_wake {
+		/* Assign GPIO24 to SIO (GPIO) for use as an interrupt source */
+		group1 {
+			pinmux = <PIO0_P24>;
```
@dzytko - Following up on this, making the following changes fixes the above performance hit and works for Pico W and Pico 2W boards (tested locally):

In `zephyr`:
```bash
--- a/boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w.dts
+++ b/boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w.dts
@@ -40,7 +40,7 @@
        airoc_wifi_host_wake: airoc_wifi_host_wake {
                /* Assign GPIO24 to SIO (GPIO) for use as an interrupt source */
                group1 {
-                       pinmux = <PIO0_P24>;
+                       pinmux = <RP2XXX_PINMUX(24, RP2_PINCTRL_GPIO_FUNC_SIO)>;
                        input-enable;
                };
        };
```

In `modules/hal/infineon`:
```
--- a/wifi-host-driver/WHD/COMPONENT_WIFI5/src/whd_chip.c
+++ b/wifi-host-driver/WHD/COMPONENT_WIFI5/src/whd_chip.c
@@ -1259,7 +1259,7 @@ static whd_result_t whd_enable_save_restore(whd_driver_t whd_driver)
         CHECK_RETURN(whd_bus_sleep(whd_driver) );
 
         /* Put SPI interface block to sleep */
-        CHECK_RETURN(whd_bus_write_register_value(whd_driver, BACKPLANE_FUNCTION, SDIO_PULL_UP, (uint8_t)1, 0xf) );
+        CHECK_RETURN(whd_bus_write_register_value(whd_driver, BACKPLANE_FUNCTION, SDIO_PULL_UP, (uint8_t)1, 0x0) );
 
         whd_driver->chip_info.save_restore_enable = WHD_TRUE;
     }
```

This produced the following thread analyzer output after 3 mins (same settings as above):

```
Thread analyze:
 thread_analyzer     : STACK: unused 560 usage 464 / 1024 (45 %); CPU: 0 %
                     : Total CPU cycles used: 530129
 rx_q[0]             : STACK: unused 1864 usage 184 / 2048 (8 %); CPU: 0 %
                     : Total CPU cycles used: 3498
 net_mgmt            : STACK: unused 584 usage 216 / 800 (27 %); CPU: 0 %
                     : Total CPU cycles used: 1146
 tcp_work            : STACK: unused 808 usage 216 / 1024 (21 %); CPU: 0 %
                     : Total CPU cycles used: 3748
 shell_uart          : STACK: unused 4488 usage 712 / 5200 (13 %); CPU: 0 %
                     : Total CPU cycles used: 2323053
 WHD                 : STACK: unused 4552 usage 568 / 5120 (11 %); CPU: 0 %
                     : Total CPU cycles used: 5511392
 airoc_event         : STACK: unused 3856 usage 240 / 4096 (5 %); CPU: 0 %
                     : Total CPU cycles used: 3203
 sysworkq            : STACK: unused 808 usage 216 / 1024 (21 %); CPU: 0 %
                     : Total CPU cycles used: 5206
 logging             : STACK: unused 448 usage 320 / 768 (41 %); CPU: 0 %
                     : Total CPU cycles used: 192782
 idle                : STACK: unused 192 usage 64 / 256 (25 %); CPU: 89 %
                     : Total CPU cycles used: 1167245897
 ISR0                : STACK: unused 1808 usage 240 / 2048 (11 %) 
```

which suggests a good path forwards.

Can I still ask if this would be a blocker for this PR? Making the above change will require some more in-depth changes to both `zephyr` and `hal_infineon` repos which would prevent people making use of networking on the Pico 2W in the meantime. I will get working in any case. Thanks, Tim

dzytko replied:
Yeah, I'd say it's a blocker. The WHD has a quite high priority of 4, so in a real application a lower priority thread might never run, which sounds like an absolute nightmare to debug. However, if others don't think that's an issue, I'm fine with merging it in its current form.

MagpieEmbedded replied:
Thanks, I'll address this change then, probably with a Kconfig option on the airoc wifi driver which can then be used here. Thanks, Tim

MagpieEmbedded commented:
@nordicjm The Kconfig files have been updated, hopefully to be more in line with your comments. Cheers, Tim

nordicjm commented on boards/raspberrypi/rpi_pico2/Kconfig.rpi_pico2:
```diff
@@ -1,5 +1,14 @@
-	select SOC_RP2350A_M33 if BOARD_RPI_PICO2_RP2350A_M33
+	select SOC_RP2350A_M33 if BOARD_RPI_PICO2_RP2350A_M33  || BOARD_RPI_PICO2_RP2350A_M33_W
+
+config MAIN_STACK_SIZE
+	int
+	default 4096
```
needs justification for this e.g. why does building hello world for this need a 4KiB stack? Irrespective of that, this symbol already exists, so would go in `Kconfig.defconfig` without the `int` and should have a condition e.g. `default 4096 if NETWORKING`

MagpieEmbedded replied:
These values were based on the `rpi_pico` Kconfig files. I've removed the `MAIN_STACK_SIZE` as it didn't seem necessary in the end, and moved the `HEAP` setting into Kconfig so that the wifi example builds as expected.

MagpieEmbedded commented:
This PR has been updated following the merge of #91227

It now efficiently runs wifi on the Pico 2W. This data was collected by using the following Kconfig settings in the `/samples/net/wifi/shell` sample while running pings for a few minutes:

```
// prj.conf
CONFIG_THREAD_ANALYZER=y
CONFIG_THREAD_ANALYZER_AUTO=y
CONFIG_THREAD_ANALYZER_AUTO_INTERVAL=10
CONFIG_LOG_BUFFER_SIZE=4096
```

```
// Serial output
Thread analyze:
 thread_analyzer     : STACK: unused 560 usage 464 / 1024 (45 %); CPU: 0 %
                     : Total CPU cycles used: 6890330
 rx_q[0]             : STACK: unused 1272 usage 776 / 2048 (37 %); CPU: 0 %
                     : Total CPU cycles used: 84694448
 net_mgmt            : STACK: unused 40 usage 760 / 800 (95 %); CPU: 0 %
                     : Total CPU cycles used: 1121439
 tcp_work            : STACK: unused 808 usage 216 / 1024 (21 %); CPU: 0 %
                     : Total CPU cycles used: 4180
 shell_uart          : STACK: unused 3648 usage 1552 / 5200 (29 %); CPU: 1 %
                     : Total CPU cycles used: 470555114
 WHD                 : STACK: unused 4552 usage 568 / 5120 (11 %); CPU: 0 %
                     : Total CPU cycles used: 59814278
 airoc_event         : STACK: unused 3856 usage 240 / 4096 (5 %); CPU: 0 %
                     : Total CPU cycles used: 7037
 sysworkq            : STACK: unused 488 usage 536 / 1024 (52 %); CPU: 0 %
                     : Total CPU cycles used: 8830705
 logging             : STACK: unused 448 usage 320 / 768 (41 %); CPU: 0 %
                     : Total CPU cycles used: 3170989
 idle                : STACK: unused 256 usage 64 / 320 (20 %); CPU: 98 %
                     : Total CPU cycles used: 43232044323
 ISR0                : STACK: unused 1808 usage 240 / 2048 (11 %)
```

@nordicjm I've also updated the Kconfig files as requested.

Thanks everyone for your help, hopefully this should be in pretty good shape to get merged in soon. Cheers, Tim

ThreeEights reviewed (dismissed):
Ready to merge.

soburi commented on boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w.dts:
```diff
@@ -0,0 +1,79 @@
+
+&pinctrl {
+	pio0_spi0_default: pio0_spi0_default {
+		/* gpio 25 is used for chip select, not assigned to the PIO */
+		group1 {
+			pinmux = <PIO0_P29>;
+		};
```
Is this "gpio 25" wrong?

ThreeEights replied:
Hmmm ... neither the comment nor the pinmux assignment are wrong, but I can see the confusion.  Long ago, the pins driven by the PIO (24 and 29) were together in the same group, and the comment was meant to inform that pin 25 was not in the group because it was used for chip select and driven directly by the SPI driver.  If you insist on clarity, the comment can be removed.

soburi replied:
Ah, I got it.

soburi commented on boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w.dts:
```diff
@@ -0,0 +1,79 @@
+	airoc_wifi_host_wake: airoc_wifi_host_wake {
+		/* Assign GPIO24 to SIO (GPIO) for use as an interrupt source */
+		group1 {
+			pinmux = <RP2XXX_PINMUX(24, RP2_PINCTRL_GPIO_FUNC_SIO)>;
```
It would be better to add this definition to include/zephyr/dt-bindings/pinctrl/rpi-pico-pinctrl-common.h.

ThreeEights replied:
Looks good!  I tested this on the Pico 2 W and Pico W.  The Pico 2 W works fine, the Pico W ... crashes in the shell:
```
[00:00:28.247,000] <err> os: ***** HARD FAULT *****
[00:00:28.247,000] <err> os: r0/a1:  0x00000008  r1/a2:  0xe000ed00  r2/a3:  0x20002efc
[00:00:28.248,000] <err> os: r3/a4:  0x00000009 r12/ip:  0x00000058 r14/lr:  0x100254ed
[00:00:28.248,000] <err> os:  xpsr:  0x20000000
[00:00:28.248,000] <err> os: Faulting instruction address (r15/pc): 0x1002f6a4
[00:00:28.248,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:28.249,000] <err> os: Current thread: 0x20000c58 (shell_uart)
[00:00:28.278,000] <err> os: Halting system
```
I suspect this has nothing to do with your changes, so I'm approving this pull request and debugging the Pico W.

MagpieEmbedded commented:
@ThreeEights @soburi I've addressed the comments in two new commits, I can squash those before merge, just separated out so it is easier to think about. Let me know what you think. Cheers, Tim

soburi commented on include/zephyr/dt-bindings/pinctrl/rpi-pico-pinctrl-common.h:
```diff
@@ -239,4 +239,35 @@
 #define USB_VBUS_DET_P25 RP2XXX_PINMUX(25, RP2_PINCTRL_GPIO_FUNC_USB)
 #define USB_VBUS_DET_P28 RP2XXX_PINMUX(28, RP2_PINCTRL_GPIO_FUNC_USB)
 
+#define SIO_P0  RP2XXX_PINMUX(0, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P1  RP2XXX_PINMUX(1, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P2  RP2XXX_PINMUX(2, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P3  RP2XXX_PINMUX(3, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P4  RP2XXX_PINMUX(4, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P5  RP2XXX_PINMUX(5, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P6  RP2XXX_PINMUX(6, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P7  RP2XXX_PINMUX(7, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P8  RP2XXX_PINMUX(8, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P9  RP2XXX_PINMUX(9, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P10 RP2XXX_PINMUX(10, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P11 RP2XXX_PINMUX(11, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P12 RP2XXX_PINMUX(12, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P13 RP2XXX_PINMUX(13, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P14 RP2XXX_PINMUX(14, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P15 RP2XXX_PINMUX(15, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P16 RP2XXX_PINMUX(16, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P17 RP2XXX_PINMUX(17, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P18 RP2XXX_PINMUX(18, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P19 RP2XXX_PINMUX(19, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P20 RP2XXX_PINMUX(20, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P21 RP2XXX_PINMUX(21, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P22 RP2XXX_PINMUX(22, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P23 RP2XXX_PINMUX(23, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P24 RP2XXX_PINMUX(24, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P25 RP2XXX_PINMUX(25, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P26 RP2XXX_PINMUX(26, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P27 RP2XXX_PINMUX(27, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P28 RP2XXX_PINMUX(28, RP2_PINCTRL_GPIO_FUNC_SIO)
+#define SIO_P29 RP2XXX_PINMUX(29, RP2_PINCTRL_GPIO_FUNC_SIO)
```
There's nothing wrong with this fix itself.
However, there are still some things that need to be done to make it perfect.

https://github.com/zephyrproject-rtos/zephyr/blob/6c569c5bface15b92ed46cf42e6fc5ee589cf2c9/include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2350-pinctrl-common.h#L70-L71

ajf58 commented on tests/drivers/gpio/gpio_api_1pin/boards/rpi_pico2_rp2350a_m33_w.overlay:
As with other cases, I'd do more like tests/drivers/gpio/gpio_api_1pin/boards/pico_plus2_rp2350b_m33.overlay , i.e. include the common base example.

MagpieEmbedded replied:
Done.

ajf58 reviewed (dismissed):
A few minor tweaks/notes for the future, but nothing blocking.

MagpieEmbedded commented on boards/raspberrypi/rpi_pico2/Kconfig.rpi_pico2:
```diff
@@ -1,5 +1,6 @@
 
 config BOARD_RPI_PICO2
-	select SOC_RP2350A_M33 if BOARD_RPI_PICO2_RP2350A_M33
+	select SOC_RP2350A_M33 if BOARD_RPI_PICO2_RP2350A_M33  || BOARD_RPI_PICO2_RP2350A_M33_W
```
Done.

MagpieEmbedded commented on boards/raspberrypi/rpi_pico2/doc/index.rst:
```diff
@@ -3,9 +3,10 @@
 ********
 
-The Raspberry Pi Pico 2 is the second-generation product in the Raspberry Pi
+The Raspberry Pi Pico 2 and Pico 2W are second-generation products in the Raspberry Pi
```
Done.

kartben commented on boards/raspberrypi/rpi_pico2/doc/index.rst:
```diff
@@ -3,9 +3,10 @@
+The Raspberry Pi Pico 2 and Pico 2W are second-generation products in the Raspberry Pi
 Pico family. From the `Raspberry Pi website <https://www.raspberrypi.com/documentation/microcontrollers/pico-series.html>`_ is referred to as Pico 2.
 
+
 There are many limitations of the board currently. Including but not limited to:
 - The Zephyr build only supports configuring the RP2350A with the Cortex-M33 cores.
```
```suggestion
 There are many limitations of the board currently. Including but not limited to:

 - The Zephyr build only supports configuring the RP2350A with the Cortex-M33 cores.
```
